catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/monitor/impl/WorkspaceServiceImpl.java

```diff
@@ -17,103 +17,142 @@
-    notNull(filterBuilder, "filterBuilder must be non-null");
     notNull(workspaceTransformer, "workspaceTransformer must be non-null");
-    notNull(filterService, "filterService must be non-null");
     notNull(securityService, "securityService must be non-null");
+    notNull(persistentStore, "persistentStore must be non-null");
+    notNull(workspaceQueryBuilder, "workspaceQueryBuilder must be non-null");
 
     this.catalogFramework = catalogFramework;
-    this.filterBuilder = filterBuilder;
     this.workspaceTransformer = workspaceTransformer;
-    this.filterService = filterService;
     this.securityService = securityService;
+    this.persistentStore = persistentStore;
+    this.maxSubscriptions = maxSubscriptions;
+    this.workspaceQueryBuilder = workspaceQueryBuilder;
+  }
+
+  @SuppressWarnings("unused" /* Needed by metatype. */)
+  public void setMaxSubscriptions(int maxSubscriptions) {
+    this.maxSubscriptions = maxSubscriptions;
   }
 
   @Override
   public String toString() {
-    return "WorkspaceServiceImpl{"
-        + "securityService="
-        + securityService
-        + ", catalogFramework="
-        + catalogFramework
-        + ", filterBuilder="
-        + filterBuilder
-        + ", workspaceTransformer="
-        + workspaceTransformer
-        + ", filterService="
-        + filterService
-        + '}';
+    return String.format(
+        "WorkspaceServiceImpl{securityService=%s, catalogFramework=%s, workspaceTransformer=%s, maxSubscriptions=%d}",
+        securityService, catalogFramework, workspaceTransformer, maxSubscriptions);
   }
 
   @Override
   public List<WorkspaceMetacardImpl> getWorkspaceMetacards() {
 
-    final QueryRequest queryRequest = createQueryRequestForAllWorkspaceMetacards();
+    final QueryRequest queryRequest = createQueryRequestForSubscribedToWorkspaces();
 
-    try {
-      return createWorkspaceMetacards(catalogFramework.query(queryRequest));
-    } catch (UnsupportedQueryException | FederationException | SourceUnavailableException e) {
-      LOGGER.warn("Error querying for workspaces", e);
+    if (queryRequest != null) {
+      try {
+        return createWorkspaceMetacards(query(queryRequest));
+      } catch (CatalogQueryException e) {
+        LOGGER.warn("Error querying for workspaces", e);
+      }
     }
 
     return Collections.emptyList();
   }
 
+  private QueryResponse query(QueryRequest queryRequest) {
+    AtomicLong hitCount = new AtomicLong(0);
+
+    QueryFunction queryFunction =
+        qr -> {
+          SourceResponse sourceResponse = catalogFramework.query(qr);
+          hitCount.compareAndSet(0, sourceResponse.getHits());
+          return sourceResponse;
+        };
+
+    ResultIterable results =
+        ResultIterable.resultIterable(queryFunction, queryRequest, maxSubscriptions);
+
+    List<Result> resultList = results.stream().collect(Collectors.toList());
+
+    long totalHits = hitCount.get();
```

Comment:
Could do this in one line: `final long totalHits = hitCount.get() > 0 ? hitCount.get() : resultList.size();`

Reply:
changed